Add support for APPENDUID response data

bitfehler opened the pull request:
If the `UIDPLUS` extension is supported, the server will reply to
`APPEND` commands with the UID of the new message. This can even be a
list of UIDs if the `MULTIAPPEND` extension is also supported.

Make this information available to the user as the result of an
`AppendCmd`. The added doc strings have links to the relevant RFCs.

Related to #131.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jonhoo/rust-imap/232)
<!-- Reviewable:end -->

bitfehler commented:
If this is of interest, I'd be happy to add `COPYUID` as well...

jonhoo commented on src/types/appended.rs:
```diff
@@ -0,0 +1,40 @@
+    }
+}
+
+impl fmt::Display for Appended {
```
Why implement `Dispay` for this type? I don't think that's actually going to be all that useful compared to the `Debug` that's already derived.

bitfehler replied:
Right, I don't remember :innocent: . I think I used it for something at some point, but definitely no longer. Will remove. Do you prefer force pushed amendments or new commits?

jonhoo replied:
New commits are generally better for changes after the first review since GitHub is _terrible_ at rendering diffs across squashes :sweat_smile:

jonhoo commented on src/client.rs:
```diff
@@ -215,7 +215,7 @@ impl<'a, T: Read + Write> AppendCmd<'a, T> {
     /// Note: be sure to set flags and optional date before you
     /// finish the command.
-    pub fn finish(&mut self) -> Result<()> {
+    pub fn finish(&mut self) -> Result<Appended> {
```
Hmm, so, this is technically a breaking change. In practice it _should_ be very rare that someone relies on `()` specifically being the return type, but it is possible. Most likely if they have a call to `finish` somewhere without a trailing `;`, meaning its return value will end up being the evaluated value of the surrounding block as well.

How about we keep the return type of `finish` as-is, mark it as deprecated, and introduce a new method (bikeshed: `send` or `complete`) that returns `Result<Appended>`? `finish` can then just be a call to that method with a `.map(|_| ())` for the return value.

jonhoo reviewed (changes requested):
Nice, thank you! And yes, `COPYUID` would be good to add, though happy to take that in a separate PR if you'd prefer.

jonhoo reviewed (changes requested):
Oh, can you please add a test for the new functionality too?

bitfehler commented:
Added a parse test much like the other one, I hope that's cool for now. I am happy to look into the integration tests, however I will be AFK for two weeks.

jonhoo commented on src/parse.rs:
```diff
@@ -698,4 +698,26 @@ mod tests {
+        // If the user has enabled UIDPLUS (RFC 4315), the response contains an APPENDUID
+        // response code followed by the UIDVALIDITY of the destination mailbox and the
+        // UID assigned to the appended message in the destination mailbox.
+        // If the MULTIAPPEND extension is also used, there can be multiple UIDs.
```
Should we then also test the multi-append case?

jonhoo commented:
Looks like there's also now a conflict in `src/parse.rs`.

bitfehler commented:
Sorry for the wait! I rebased on current master and added another test for the MULTIAPPEND response. I hope this is fine for this PR. I'll start looking into the integration tests to see if I can add some for all the stuff I added now.

jonhoo reviewed (approved):
Looks great, thanks!
